File: backend/src/main/kotlin/at/wrk/tafel/admin/backend/common/http/ContentDispositionUtil.kt
package at.wrk.tafel.admin.backend.common.http

import org.springframework.http.ContentDisposition
import org.springframework.http.HttpHeaders
import java.nio.charset.StandardCharsets

/**
 * Builds the `Content-Disposition` header for a file download/export response. A raw
 * `"inline; filename=$name"`/`"attachment; filename=$name"` string is unsafe: a stored filename
 * containing `;`, `"` or non-ASCII characters produces either a malformed header or a spoofed
 * download name (see issue #3438). Spring's [ContentDisposition] quotes the filename and adds an
 * RFC 5987-encoded `filename*` parameter instead, so every character survives round-trip
 * regardless of what the stored filename contains.
 */
object ContentDispositionUtil {

    fun inline(filename: String): HttpHeaders = headers(ContentDisposition.inline(), filename)

    fun attachment(filename: String): HttpHeaders = headers(ContentDisposition.attachment(), filename)

    private fun headers(builder: ContentDisposition.Builder, filename: String): HttpHeaders {
        val headers = HttpHeaders()
        headers.contentDisposition = builder.filename(filename, StandardCharsets.UTF_8).build()
        return headers
    }
}

File: frontend/src/main/webapp/src/app/common/util/content-disposition.util.ts
/**
 * Extracts the filename from a `Content-Disposition` header value, preferring the RFC 5987
 * `filename*=<charset>''<percent-encoded>` parameter when present - the backend always sets it
 * alongside a plain `filename="..."` fallback (`ContentDispositionUtil`), and it round-trips
 * every character regardless of what the stored filename contains. A naive
 * `header.split('filename')[1].split('=')[1]` breaks against that quoted/dual-parameter shape (it
 * picks up the plain parameter's surrounding quotes literally), which is what happened here -
 * see issue #3438.
 */
export function parseContentDispositionFilename(header: string): string {
  const extended = header.match(/filename\*=[^']*''([^;]+)/i);
  if (extended) {
    return decodeURIComponent(extended[1].trim());
  }

  const plain = header.match(/filename="?([^";]+)"?/i);
  return plain ? plain[1].trim() : header.trim();
}
